### PR TITLE
Getters: Ingress Global, Generators: Ingress Global, Service-route

### LINKED ISSUE #1161: Add ingress-global getter and generator to space-cli
We've added a new resource type `ingress-global` in SpaceCloud. Need to make sure it is present in `space-cli` as well. Specially in `space-cli get all`.

>If you want this feature to be implemented, give it a thumbs up reaction, so that we can determine which features are important to you.
>👍

### PR DESCRIPTION
**Pull request by KavishShah09**
fixes #1161

### CODE REVIEW

File: space-cli/cmd/modules/all.go [hunk 1]
```diff
@@ -197,6 +197,14 @@ func getAllProjects(cmd *cobra.Command, args []string) error {
 		return nil
 	}
 
+	objs, err = ingress.GetIngressGlobal(projectName, "ingress-global")
```
Comment (YourTechBud) [inline space-cli/cmd/modules/all.go:200]: This should be before `ingress-routes`

Comment (YourTechBud) [reply space-cli/cmd/modules/all.go:200]: Also, `db-prepared-query` should be after `db-schemas`

File: gateway/server/handlers/config_routing.go [hunk 2]
```diff
@@ -130,18 +130,13 @@ func HandleDeleteProjectRoute(adminMan *admin.Manager, syncMan *syncman.Manager)
...
-
 	return func(w http.ResponseWriter, r *http.Request) {
 		// Extract the required path parameters
 		vars := mux.Vars(r)
 		projectID := vars["project"]
 
-		// Get request body
-		req := new(request)
-		_ = json.NewDecoder(r.Body).Decode(req)
+    config := new(config.GlobalRoutesConfig)
```
Comment (YourTechBud) [inline gateway/server/handlers/config_routing.go:138]: Indent

File: space-cli/cmd/modules/database/getters.go [hunk 3]
```diff
@@ -120,11 +120,7 @@ func GetDbPreparedQuery(project, commandName string, params map[string]string) (
 	for _, item := range payload.Result {
 		obj := item.(map[string]interface{})
 		meta := map[string]string{"project": project, "db": obj["db"].(string), "id": obj["id"].(string)}
```
Comment (YourTechBud) [inline space-cli/cmd/modules/database/getters.go:122]: delete the fields `db`, `id` from obj

File: space-cli/cmd/modules/ingress/commands.go [hunk 4]
```diff
@@ -16,7 +16,14 @@ func GenerateSubCommands() []*cobra.Command {
 		Example: "space-cli generate ingress-route config.yaml --project myproject --log-level info",
 	}
 
-	return []*cobra.Command{generateroutes}
+	var generateIngressGlobal = &cobra.Command{
+		Use:     "ingress-global [path to config file]",
+		RunE:    actionGenerateIngressGlobal,
+		Aliases: []string{"ingress-global"},
```
Comment (YourTechBud) [inline space-cli/cmd/modules/ingress/commands.go:22]: Not needed

File: space-cli/cmd/modules/ingress/commands.go [hunk 5]
```diff
@@ -71,3 +83,35 @@ func actionGenerateIngressRouting(cmd *cobra.Command, args []string) error {
 
 	return utils.AppendConfigToDisk(dbrule, dbruleConfigFile)
 }
+
+func actionGenerateIngressGlobal(cmd *cobra.Command, args []string) error {
+	if len(args) != 1 {
+		return utils.LogError("incorrect number of arguments. Use -h to check usage instructions", nil)
```
Comment (YourTechBud) [inline space-cli/cmd/modules/ingress/commands.go:89]:
```suggestion
		return utils.LogError("Incorrect number of arguments. Use -h to check usage instructions", nil)
```

File: space-cli/cmd/modules/ingress/commands.go [hunk 6]
```diff
@@ -71,3 +83,35 @@ func actionGenerateIngressRouting(cmd *cobra.Command, args []string) error {
...
+	return utils.AppendConfigToDisk(dbrule, dbruleConfigFile)
+}
+
+func actionGetIngressGlobal(cmd *cobra.Command, args []string) error {
+	// Get the project and url parameters
+	project, check := utils.GetProjectID()
+	if !check {
+		return utils.LogError("Project not specified in flag", nil)
+	}
+	commandName := "ingress-global"
```
Comment (YourTechBud) [inline space-cli/cmd/modules/ingress/commands.go:106]: No need to make a variable

File: space-cli/cmd/modules/ingress/generate.go [hunk 7]
```diff
@@ -105,3 +105,45 @@ func generateIngressRouting() (*model.SpecObject, error) {
...
+
+func generateIngressGlobal() (*model.SpecObject, error) {
+
+	project := ""
+	if err := input.Survey.AskOne(&survey.Input{Message: "Enter Project:"}, &project); err != nil {
+		return nil, err
+	}
+
+	var headers model.Header
+	if err := input.Survey.AskOne(&survey.Input{Message: "Enter Request Header Key:"}, &headers.Key); err != nil {
```
Comment (YourTechBud) [inline space-cli/cmd/modules/ingress/generate.go:117]: Add default value `Foo`

File: space-cli/cmd/modules/ingress/generate.go [hunk 8]
```diff
@@ -105,3 +105,45 @@ func generateIngressRouting() (*model.SpecObject, error) {
...
+	}
+	if err := input.Survey.AskOne(&survey.Input{Message: "Enter Request Header Value:"}, &headers.Value); err != nil {
+		return nil, err
+	}
+	if err := input.Survey.AskOne(&survey.Input{Message: "Enter Request Header Option:"}, &headers.Op); err != nil {
+		return nil, err
+	}
+
+	var resHeaders model.Header
+	if err := input.Survey.AskOne(&survey.Input{Message: "Enter Response Header Key:"}, &resHeaders.Key); err != nil {
```
Comment (YourTechBud) [inline space-cli/cmd/modules/ingress/generate.go:128]: Add default value `Foo`

File: space-cli/cmd/modules/ingress/generate.go [hunk 9]
```diff
@@ -105,3 +105,45 @@ func generateIngressRouting() (*model.SpecObject, error) {
...
+	project := ""
+	if err := input.Survey.AskOne(&survey.Input{Message: "Enter Project:"}, &project); err != nil {
+		return nil, err
+	}
+
+	var headers model.Header
+	if err := input.Survey.AskOne(&survey.Input{Message: "Enter Request Header Key:"}, &headers.Key); err != nil {
+		return nil, err
+	}
+	if err := input.Survey.AskOne(&survey.Input{Message: "Enter Request Header Value:"}, &headers.Value); err != nil {
```
Comment (YourTechBud) [inline space-cli/cmd/modules/ingress/generate.go:120]: Add default value `Bar`

File: space-cli/cmd/modules/ingress/generate.go [hunk 10]
```diff
@@ -105,3 +105,45 @@ func generateIngressRouting() (*model.SpecObject, error) {
...
+	}
+	if err := input.Survey.AskOne(&survey.Input{Message: "Enter Request Header Option:"}, &headers.Op); err != nil {
+		return nil, err
+	}
+
+	var resHeaders model.Header
+	if err := input.Survey.AskOne(&survey.Input{Message: "Enter Response Header Key:"}, &resHeaders.Key); err != nil {
+		return nil, err
+	}
+	if err := input.Survey.AskOne(&survey.Input{Message: "Enter Response Header Value:"}, &resHeaders.Value); err != nil {
```
Comment (YourTechBud) [inline space-cli/cmd/modules/ingress/generate.go:131]: Add default value `Bar`

File: space-cli/cmd/modules/services/generate.go [hunk 11]
```diff
@@ -134,3 +134,60 @@ func GenerateService(projectID, dockerImage string) (*model.SpecObject, error) {
...
+		},
+		Spec: map[string]interface{}{
+			"routes": map[string]interface{}{
+				"source": map[string]interface{}{
+					"port": port,
+				},
+				"targets": map[string]interface{}{
+					"type":    routeType,
+					"version": version,
+					"weight":  weight,
```
Comment (YourTechBud) [inline space-cli/cmd/modules/services/generate.go:186]: Hard code weight to `100`

File: space-cli/cmd/modules/services/generate.go [hunk 12]
```diff
@@ -134,3 +134,60 @@ func GenerateService(projectID, dockerImage string) (*model.SpecObject, error) {
...
+
+	version := ""
+	if err := input.Survey.AskOne(&survey.Input{Message: "Enter Version:"}, &version); err != nil {
+		return nil, err
+	}
+
+	weight := 0
+	if err := input.Survey.AskOne(&survey.Input{Message: "Enter Weight:"}, &weight); err != nil {
+		return nil, err
+	}
```
Comment (YourTechBud) [inline space-cli/cmd/modules/services/generate.go:169]: Don't ask

File: space-cli/cmd/modules/services/generate.go [hunk 13]
```diff
@@ -134,3 +134,60 @@ func GenerateService(projectID, dockerImage string) (*model.SpecObject, error) {
...
+			"id":      id,
+			"project": projectID,
+		},
+		Spec: map[string]interface{}{
+			"routes": map[string]interface{}{
+				"source": map[string]interface{}{
+					"port": port,
+				},
+				"targets": map[string]interface{}{
+					"type":    routeType,
```
Comment (YourTechBud) [inline space-cli/cmd/modules/services/generate.go:184]: hard code to `internal`

File: space-cli/cmd/model/service.go [hunk 14]
```diff
@@ -147,3 +147,13 @@ type Response struct {
...
+
+// Header describes the operation to be performed on the header
+type Header struct {
+	Key   string `json:"key" yaml:"key"`
+	Value string `json:"value" yaml:"value"`
+	Op    string `json:"op" yaml:"op"`
+}
+
+// Headers describes an array of headers
+type Headers []Header
```
Comment (YourTechBud) [inline space-cli/cmd/model/service.go:159]: Can we do without this?

File: space-cli/cmd/modules/all.go [hunk 15]
```diff
@@ -85,6 +85,14 @@ func getAllProjects(cmd *cobra.Command, args []string) error {
 		return nil
 	}
 
+	objs, err = database.GetDbPreparedQuery(projectName, "db-prepared-query", map[string]string{})
+	if err != nil {
+		return nil
+	}
+	if err := createConfigFile("18", "db-prepared-query", objs); err != nil {
```
Comment (YourTechBud) [inline space-cli/cmd/modules/all.go:92]: Keep the file name in order. This should be 05

File: space-cli/cmd/modules/all.go [hunk 16]
```diff
@@ -165,35 +173,35 @@ func getAllProjects(cmd *cobra.Command, args []string) error {
 		return nil
 	}
 
-	objs, err = ingress.GetIngressRoutes(projectName, "ingress-routes", map[string]string{})
+	objs, err = ingress.GetIngressGlobal(projectName, "ingress-global")
 	if err != nil {
 		return nil
 	}
-	if err := createConfigFile("15", "ingress-routes", objs); err != nil {
+	if err := createConfigFile("19", "ingress-global", objs); err != nil {
```
Comment (YourTechBud) [inline space-cli/cmd/modules/all.go:180]: Keep the file name in order

File: space-cli/cmd/modules/ingress/generate.go [hunk 17]
```diff
@@ -105,3 +105,49 @@ func generateIngressRouting() (*model.SpecObject, error) {
...
+	headersKey := ""
+	if err := input.Survey.AskOne(&survey.Input{Message: "Enter Request Header Key:", Default: "Foo"}, &headersKey); err != nil {
+		return nil, err
+	}
+	headersValue := ""
+	if err := input.Survey.AskOne(&survey.Input{Message: "Enter Request Header Value:", Default: "Bar"}, &headersValue); err != nil {
+		return nil, err
+	}
+	headersOption := ""
+	if err := input.Survey.AskOne(&survey.Input{Message: "Enter Request Header Option:"}, &headersOption); err != nil {
```
Comment (YourTechBud) [inline space-cli/cmd/modules/ingress/generate.go:125]: The `op` stands for operation and not option. I meant the user should get a select or something. Three possible values, set, add, del

File: space-cli/cmd/modules/services/generate.go [hunk 18]
```diff
@@ -134,3 +134,50 @@ func GenerateService(projectID, dockerImage string) (*model.SpecObject, error) {
...
+		Meta: map[string]string{
+			"id":      id,
+			"project": projectID,
+		},
+		Spec: map[string]interface{}{
+			"routes": map[string]interface{}{
+				"source": map[string]interface{}{
+					"port": port,
+				},
+				"targets": map[string]interface{}{
```
Comment (YourTechBud) [inline space-cli/cmd/modules/services/generate.go:173]: I think targets is an array of objects

File: space-cli/cmd/modules/services/generate.go [hunk 19]
```diff
@@ -134,3 +134,50 @@ func GenerateService(projectID, dockerImage string) (*model.SpecObject, error) {
...
+
+	v := &model.SpecObject{
+		API:  "/v1/runner/{project}/service-routes/{id}",
+		Type: "service-route",
+		Meta: map[string]string{
+			"id":      id,
+			"project": projectID,
+		},
+		Spec: map[string]interface{}{
+			"routes": map[string]interface{}{
```
Comment (YourTechBud) [inline space-cli/cmd/modules/services/generate.go:169]: This is an array of objects. Can you confirm

Comment (KavishShah09) [reply space-cli/cmd/modules/services/generate.go:169]: Yes it is

File: space-cli/cmd/modules/services/generate.go (hunk 18 above)
Comment (KavishShah09) [reply space-cli/cmd/modules/services/generate.go:173]: Yes it is

File: space-cli/cmd/modules/services/generate.go (hunk 19 above)
Comment (YourTechBud) [reply space-cli/cmd/modules/services/generate.go:169]: So lets make it an array here

Comment (KavishShah09) [reply space-cli/cmd/modules/services/generate.go]: Okay

File: space-cli/cmd/modules/services/generate.go (hunk 18 above)
Comment (KavishShah09) [reply space-cli/cmd/modules/services/generate.go]: Okay

**Pull request merged**